### DISCUSSION TITLE
Flashing Firmware Tips (OctoPi)

### DISCUSSION THREAD

**Discussion by banhathome** [bassamanator/Sovol-SV06-firmware#78] [Show and tell]
Hi everyone! I just set up Klipper on my new SV06 and I just wanted to share some random tips on how I managed to get it working.

I used a Raspberry Pi 4 Model B with OctoPi installed. Klipper actually uses OctoPi as the basis for their [installation guide](https://www.klipper3d.org/Installation.html). I recommend going through this first, then using the rest of this post as a troubleshooting guide for getting it working on SV06 specifically.

## Main Tips

* Make sure you actually download Klipper onto your Raspberry Pi! 

  ```zsh
  git clone https://github.com/Klipper3d/klipper
  ./klipper/scripts/install-octopi.sh
  ```

* I was able to flash the Klipper firmware using the SD card that came with my SV06! However, the firmware provided in this repo DID NOT work for me. 
  * I had to compile the firmware myself:
    ```zsh
    cd ~/klipper/
    make menuconfig
    ```
  * In this menu:
    * `Enable extra low-level configuration options`: checked
    * `Micro-controller Architecture`: STMicroelectronics STM32
    * `Processor model`: STM32F103
    * `Disable SWD at startup`: checked
    * `Bootloader offset`: 28KiB bootloader
    * `Communication interface`: Serial (on USART1 PA10/PA9)
    * `Baud rate for serial port`: 115200

    ![Screenshot 2023-07-25 at 10 01 48 PM](https://github.com/bassamanator/Sovol-SV06-firmware/assets/8410793/d244ede4-a677-4bb5-8b02-185257858498)

  * Then press Q and save.
  * Then enter `make` in the terminal and it generate a `klipper.bin` file (`~/klipper/out/klipper.bin`) which you can now use to flash onto your printer.
* Flashing Firmware Tips:
  * Make sure you have your Finder or File Explorer to show hidden files because they should not be on your SD card when you flash
    * This is ESPECIALLY true for MacOS!!!! Spotlight loves filling up your drive with random hidden folders! You might even need to configure Spotlight to ignore your drive because it might add in the folder right before you eject the SD card.
  * The screen will remain blank even after you've successfully flashed the firmware! The screen will appear after you plug in your Raspberry Pi and Klipper successfully starts. 

* The `printer.cfg` MUST reside in your home directory (NOT `~/printer_data/config` directory 🙅‍♂️). In other words, you can follow the instructions in the README of this repo, but then type `mv ~/printer_data/config/printer.cfg ~/printer_data/config/moonraker.cfg ~/printer_data/config/cfgs ~` to move the relevant config files to your home directory.
* Make sure you update `moonraker.cfg` and `cfgs/misc_macros.cfg` to replace `pi` with your username if you changed it.
* For some reason, the `by-id` mcu serial path did not work for me, so instead just follow the same instructions but use the `by-path` instead.
* To get OctoPrint talking to the Klipper service, make sure you follow [these instructions in Klipper's guide](https://www.klipper3d.org/Installation.html#configuring-octoprint-to-use-klipper).
* You can `firmware_restart` via the OctoPrint terminal as long as your Raspberry Pi with Klipper running on it is connected to your SV06 with the Klipper firmware successfully flashed on it.

## Extras:

* You can view your bed mesh in OctoPrint. "Settings" > "Plugin Manager". Then click on the "+ Get More" button and search for "Bed Level Visualizer". 
* If you're using Cura: I pretty much deleted my entire Start G-Code and End G-Code because it basically doesn't even work anymore with Klipper. My start code is just `PRINT_START BED=...` and `PURGE_LINE`, and my end code is `PRINT_END`.
* Don't forget to set your Pressure Advance! The config in this repo doesn't have it defined because you should determine it for yourself since every printer is different. Guide: https://github.com/Klipper3d/klipper/blob/master/docs/Pressure_Advance.md
* Install the "M73 Progress Plugin" in OctoPrint so that your SV06 screen displays the progress of your print.
---

I hope this helps!!! I'll be editing this to add more info as I think of it, and feel free to drop any questions in the comments and I'll get back to you ASAP

**Comment by bassamanator** [comment]
@chrisngobanh Thanks for this! I'm sure Octopi/Octoprint users will find this useful.

I guess you opted for Octoprint over Mainsail/Fluidd because of the many plugins available for Octoprint?

**Reply by banhathome** [reply]
Klipper was actually an afterthought for me haha. I was already tinkering with my printer a lot so I thought it'd be a good idea to get Klipper on it to expand what I can mess around with.

**Comment by brogrene** [comment]
Thanks for the helpful guide towards getting me going with Klipper! I am just getting into 3D printing on this SV06 and just got OctoPi running smoothly so didn't want to redo everything. I'd contacted Sovol and they have no plans to add newer featured Marlin firmware but I didn't want to invest another $100 more into their recent Klipper Screen.  
Overall, a bit of trouble with understanding transferring files to Pi but eventually got it figured using WinSCP gui instead of trying to learn more Linux. Also, ran into trouble with OctoPi using 115200. Seems it needed to set to 250000

Since this was put out in July have you made any further improvements/suggestions to your "OctoPi Klipperized SV06"? I'm interested in getting the SV06 LCD screen menu to have more similar features with Sovol's firmware.  I liked their "Probe Z Offset" sequence with turning the knob to adjust it instead of going back over to OctoPi and typing in terminal commands. I've also gotten used to how their filament load/unload menu sequence moves extruder out of the way to purge. Oh, and the OctoPi plugin for Bed Visualizer doesn't seem to work the same as it did with Marlin when updating the mesh. Probably needs different gcode entries?

**Reply by banhathome** [reply]
Yeah it's a little annoying having to use the OctoPrint terminal for everything instead of using the screen. You can customize the LCD menu yourself if you're savvy enough with Klipper. I did some customization for my printer but nothing really interesting enough to share atm, but at some point I'd love to include some of my scripts in this repo
